### PR TITLE
feat: JSON Summary Reports, Coverage Comparison & CI Enforcement

### LINKED ISSUE #47: JSON coverage summary output
I'd like to output a coverage summary in JSON similar to [`jest --coverage --coverageReporters="json-summary"`](https://stackoverflow.com/questions/60967561/jest-coverage-how-can-i-get-a-total-percentage-of-coverage/60968723#60968723) ([relevant docs](https://jestjs.io/docs/configuration#coveragereporters-arraystring--string-options)) to produce 
```json
  "total": {
    "lines": { "total": 21777, "covered": 65, "skipped": 0, "pct": 0.3 },
    "statements": { "total": 24163, "covered": 72, "skipped": 0, "pct": 0.3 },
    "functions": { "total": 5451, "covered": 16, "skipped": 0, "pct": 0.29 },
    "branches": { "total": 6178, "covered": 10, "skipped": 0, "pct": 0.16 }
  }
```

I'm torn on whether this belongs in LocalCoverage.jl or Coverage.jl. One the one hand ([from a familiar thread](https://discourse.julialang.org/t/local-gui-coverage/6842/9)):
```julia
using Coverage
coverage = process_folder()
LCOV.writefile("lcov.info", coverage)
```
and note the other possible values of that `jest` argument are `["clover", "json", "lcov", "text"]`, so you could argue that lcov and json should be an alternative output option of Coverage.jl.

But on the other hand, using `jest --coverageReporters="text-summary"` produces:
```
=============================== Coverage summary ===============================
Statements   : 100% ( 166/166 )
Branches     : 75% ( 18/24 )
Functions    : 100% ( 49/49 )
Lines        : 100% ( 161/161 )
================================================================================
```
which is very similar to the text summary output of LocalCoverage.jl:
```
Coverage of /Users/haiiro/NoSync/LocalCoverage.jl/test/DummyPackage/
┌─────────────────────┬───────┬─────┬──────┬──────┐
│ Filename            │ Lines │ Hit │ Miss │    % │
├─────────────────────┼───────┼─────┼──────┼──────┤
│ src/DummyPackage.jl │     1 │   1 │    0 │ 100% │
│ src/bar.jl          │     2 │   1 │    1 │  50% │
│ src/corge/corge.jl  │     1 │   1 │    0 │ 100% │
│ src/corge/grault.jl │     1 │   0 │    1 │   0% │
│ src/qux.jl          │     2 │   1 │    1 │  50% │
│ src/qux.jl          │     2 │   1 │    1 │  50% │
├─────────────────────┼───────┼─────┼──────┼──────┤
│ TOTAL               │     9 │   5 │    4 │  56% │
└─────────────────────┴───────┴─────┴──────┴──────┘
```

Personally, I think it should go here since the necessary components are already conveniently located and need only be put in a dict to make for easy json output.

For background, I'm interested in storing the json during CI, then reading it during subsequent runs in order to provide a pass/fail based on coverage increase/decrease, similar to what you get out of the box with codecov.io. (It's either this or cough up the $500/yr for codecov pro!)

Happy to PR.

### PR DESCRIPTION
**Pull request by twilsonco**
This PR introduces a suite of robust coverage reporting, comparison, and continuous integration (CI) enforcement capabilities to `LocalCoverage.jl`. These additions allow developers and automated CI pipelines to generate Jest-compatible JSON summaries, compare current coverage with baseline summaries, display coverage deltas in a clear table, and fail builds or scripts if overall coverage decreases or falls below a target percentage threshold.

This PR addresses all requirements for coverage summary reporting.
---

### 1. Jest-Compatible JSON Coverage Summary
* Added a new function `generate_json_summary(coverage::PackageCoverage, filename = "coverage-summary.json"; test_args = [""])` which outputs a coverage report mimicking the structure of Jest's `json-summary` reporter (tracking lines, statements, functions, and branches).
* Fully integrated JSON generation into the main `generate_coverage` pipeline via keywords:
  - `json_summary::Bool = false` (automatically set to `true` if comparison is requested).
  - `json_summary_filename::String = "coverage-summary.json"`.
* Supports replacing the default top-level `"total"` key in the summary with customized test set names derived from `test_args`.

### 2. Coverage Comparison & Delta Reporting
* Added `compare_coverage_json_summaries(old::AbstractString, new::AbstractString, io::IO = stdout)` to parse and compare coverage summaries (from either JSON strings or files).
* Generates a beautifully formatted terminal/CI table using `PrettyTables` detailing:
  - Each individual file and the overall package.
  - Old coverage % vs. New coverage %.
  - Precise delta percentages (e.g. `+8.57%` or `-2.30%`).
* Supports **automatic pass gracefully** if no prior baseline file exists, ensuring seamless bootstrapping of CI/CD loops.
* Integrated strict enforcement in `generate_coverage`:
  - `json_comparison_summary_filename::Union{Nothing,String} = nothing` specifies the baseline JSON path to compare against.
  - `json_summary_comparison_fail_on_decrease::Bool = false` determines whether to throw an error and halt execution if overall coverage has dropped.

### 3. CI-Friendly CLI Enforcement (`report_coverage_and_exit`)
* Added `report_coverage_and_exit` to facilitate CLI/CI automation.
* This method prints the coverage summary and exits with status code `0` if the target coverage threshold is met, or status code `1` otherwise.
* Can be invoked directly from command-line scripts:
  ```bash
  julia --project=@. -e 'using LocalCoverage; report_coverage_and_exit(; target_coverage=90)'
  ```

### 4. PrettyTables 3.0 Compatibility
* Implemented compile-time compatibility branching (`@static if pkgversion(PrettyTables) < v"3.0.0"`) to support both older PrettyTables releases and the modern v3.0 API. This includes adaptive usage of `Highlighter`/`TextHighlighter`, tabular keywords, layout configurations, and text format styling.

---

## File Changes & Coverage Analysis

* **`Project.toml`**: Added dependency on `JSON` with appropriate `[compat]` bounds.
* **`README.md`**: Extensively documented the new functions, inputs, options, CLI usage, and CI integration, complete with layout previews.
* **`src/LocalCoverage.jl`**: Refactored the core logic to return the `PackageCoverage` object on success, implemented XML generation utilities, added `generate_json_summary`, comparison routines, and `report_coverage_and_exit`.
* **`test/Project.toml`**: Added test-environment dependencies.
* **`test/runtests.jl`**: Added comprehensive test coverage (~200 lines of robust tests) to validate JSON parsing/saving, file comparisons, CLI subprocess exit codes, edge cases, missing file paths, and Aqua quality assurance checks.

---

### Generating a JSON summary:
```julia
using LocalCoverage
generate_coverage(pkg; json_summary = true)
```

### CI check failing on coverage decrease:
```julia
using LocalCoverage
generate_coverage(pkg; 
                  json_comparison_summary_filename = "coverage/previous-summary.json",
                  json_summary_comparison_fail_on_decrease = true)
```

### CI check enforcing a 90% threshold:
```bash
julia --project=@. -e 'using LocalCoverage; report_coverage_and_exit(target_coverage=90)'
```

---

## Verification & Testing

Tests pass locally with 100% success, confirming:
1. Valid JSON format structures (matching the Jest standard).
2. Proper detection of coverage increases (passing), unchanged metrics (passing), and coverage decreases (raising errors when requested).
3. CLI exit codes matching the coverage thresholds across mock test scenarios.

closes #47

### CODE REVIEW

**Comment by jakobjpeters** [comment]
Cool! The JSON schema should be documented so that users know what to expect and this package is semantically versioned to support that. I like documentating JSON similar to this:

````
- `a`: description...
- `b`:
- `c`:
- `d`:
- `e`: an `Object2` used for ...

### Object1
```json
{
    "a": number
    "b": array
    "c": null
    "d": string
    "e": object
}
```

### Object2

```json
{ ... }
```
````

**Comment by twilsonco** [comment]
@jakobjpeters I don't entirely understand, but how's this (from the updated readme)?

---

#### JSON Summary Schema

The JSON summary mimics [Jest's standard `json-summary` reporter format](https://stackoverflow.com/a/60968723) ([relevant section of Jest docs](https://jestjs.io/docs/configuration#coveragereporters-arraystring--string-options)). It contains a top-level key for the package total (or the name of the test set if customized), followed by keys for each individual tracked file path.

Each section contains sub-metrics for `lines`, `statements`, `functions`, and `branches` with the following structure:
* `total`: Total number of lines tracked.
* `covered`: Number of lines hit.
* `skipped`: Always `0` (included for compatibility).
* `pct`: Coverage percentage (as a float).

```json
{
  "total": {
    "lines": { "total": 10, "covered": 5, "skipped": 0, "pct": 50.0 },
    "statements": { "total": 10, "covered": 5, "skipped": 0, "pct": 50.0 },
    "functions": { "total": 10, "covered": 5, "skipped": 0, "pct": 50.0 },
    "branches": { "total": 10, "covered": 5, "skipped": 0, "pct": 50.0 }
  },
  "src/bar.jl": {
    "lines": { "total": 5, "covered": 2, "skipped": 0, "pct": 40.0 },
    "statements": { "total": 5, "covered": 2, "skipped": 0, "pct": 40.0 },
    "functions": { "total": 5, "covered": 2, "skipped": 0, "pct": 40.0 },
    "branches": { "total": 5, "covered": 2, "skipped": 0, "pct": 40.0 }
  }
}
```

**Comment by jakobjpeters** [comment]
That's great, thank you! I think it needs to be part of the docstring too. Here's an example, if it's helpful, which provides more details for each field. I'm not set on the structure of it though so feel free to adapt if it suits you :)

````
"""
$(SIGNATURES)

Generate a coverage JSON summary in the package `coverage` directory, mimicking the structure
of the Jest `json-summary` reporter.

If `test_args` are provided, the top level `"total"` key is replaced with the name of the test set.

# JSON Schema

The JSON summary implements [Jest's standard json-summary reporter format](https://stackoverflow.com/a/60968723) ([relevant section of Jest docs](https://jestjs.io/docs/configuration#coveragereporters-arraystring--string-options)).

The top-level object contains a variable number of keys, including the package `total` (or the name of the test set, if customized) followed by keys for each individual tracked file path.

- top-level object
    - 'total`: the combined results of all other keys
    - `file`: the results for the specified file

```json
{
  "total": node,
  "src/foo.jl": node
}
```

- `node`
    - `lines`: ?
    - `statements`: ?
    - `functions`: ?
    - `branches`: ?

```json
{
  "lines": data,
  "statements": data,
  "functions": data,
  "branches": data  
}
```

- `data`
    - `total`: the number of identified nodes
    - `covered: the number of nodes ran by tests
    - `skipped`: the number of nodes not ran in tests
    - `pct`: the percentage of covered nodes `100 * covered / total`

```json
{
  "total": number,
  "covered": number,
  "skipped": number,
  "pct": number
}
```
"""

File: src/LocalCoverage.jl [hunk 1]
```diff
@@ -238,6 +240,14 @@ step is skipped allowing an easier use in combination with other test packages.
 a test is run on files/folders that *are not* in the list, then those files will be
 shown as having 0% coverage.
 
+- `json_summary = false` determines whether a JSON summary of coverage is generated.
```
Comment (tpapp) [inline src/LocalCoverage.jl]: I think that this function has too many keyword arguments now (not this PR's fault, it is just following existing style, but this may be a good time to fix things).

I am wondering if it would make sense to have a 
```julia
json_summary::Union{Nothing,JSONSummaryOptions} = nothing
```
default, with the alternative
```julia
Base.@kwdef struct JSONSummaryOptions
    flename = "coverage-summary.json"
    comparison_filename = nothing
    comparison_fail_on_decrease = false
end
```
This is just a sketch, typed in the browser, names up to bikeshedding, etc.

**Review by tpapp — COMMENTED** [review]
Thanks for this very nice PR! I especially appreciate the thorough tests and documentation. I am very busy now so I just made a quick comment, I will play around with it in detail. If you do not hear from me in a week, feel free to ping, I don't want this to get forgotten.

File: src/LocalCoverage.jl (hunk 1 above)
Comment (twilsonco) [reply src/LocalCoverage.jl]: I agree. Great idea to simplify the function signature. I'll implement this.

File: src/LocalCoverage.jl [hunk 2]
```diff
@@ -734,6 +844,114 @@ Get the percentage of lines covered in the total, with formatting
...
+        println(io, "No previous coverage summary found. Automatic pass.")
+        return true
+    end
+    old_data = if is_json(old)
+        JSON.parse(old; dicttype=OrderedDict{String,Any})
+    elseif isfile(old)
+        JSON.parsefile(old; dicttype=OrderedDict{String,Any})
+    else
+        JSON.parse(old; dicttype=OrderedDict{String,Any})
+    end
```
Comment (tpapp) [inline src/LocalCoverage.jl]: The same kind of logic is implemented twice, perhaps make into a function?

File: src/LocalCoverage.jl [hunk 3]
```diff
@@ -734,6 +844,114 @@ Get the percentage of lines covered in the total, with formatting
...
+        end
+        return "total" # fallback
+    end
+    
+    old_key = find_overall_key(old_data)
+    new_key = find_overall_key(new_data)
+    
+    # Extract overall coverage percentages
+    old_pct = haskey(old_data, old_key) ? old_data[old_key]["lines"]["pct"] : nothing
+    new_pct = haskey(new_data, new_key) ? new_data[new_key]["lines"]["pct"] : nothing
```
Comment (tpapp) [inline src/LocalCoverage.jl]: Again, repeated logic, maybe move into a mini-function.

File: src/LocalCoverage.jl [hunk 4]
```diff
@@ -734,6 +844,114 @@ Get the percentage of lines covered in the total, with formatting
...
+    
+    # Get all keys to compare
+    keys_all = collect(keys(new_data))
```
Comment (tpapp) [inline src/LocalCoverage.jl]: I wonder if it would make sense to use eg `Set` for keys in the code, then this is simply a `union` operation. Works without `Set` too though. Does the order of keys matter?

File: src/LocalCoverage.jl [hunk 5]
```diff
@@ -734,6 +844,114 @@ Get the percentage of lines covered in the total, with formatting
...
+        new_pct_str = new_val isa Number ? "$new_val%" : new_val
+        
+        rows[i, 1] = k
+        rows[i, 2] = old_pct_str
+        rows[i, 3] = new_pct_str
+        rows[i, 4] = delta_str
+    end
+    
+    # Print the table using version-appropriate kwargs
+    @static if pkgversion(PrettyTables) < v"3.0.0"
```
Comment (tpapp) [inline src/LocalCoverage.jl:962]: Now that we are doing this in two places, it would make sense to extract into a function.

**Review by tpapp — CHANGES_REQUESTED** [review]
Further minor changes, mainly about code style.

If you address these and the previous suggestions, I am happy to merge this, it would be a great feature.

File: src/LocalCoverage.jl [hunk 6]
```diff
@@ -449,6 +482,92 @@ function generate_xml(coverage::PackageCoverage, filename="cov.xml")
     @info("generated cobertura XML $(joinpath(coverage.package_dir, COVDIR, filename)).")
 end
 
+"""
+$(SIGNATURES)
+
+Generate a coverage JSON summary in the package `coverage` directory, mimicking the structure
+of the Jest `json-summary` reporter.
```
Comment (tpapp) [inline src/LocalCoverage.jl]:
```suggestion
of the [Jest](https://jestjs.io/) `json-summary` reporter.
```
We should probably add a link. (is this the right one?)

File: src/LocalCoverage.jl [hunk 7]
```diff
@@ -449,6 +482,92 @@ function generate_xml(coverage::PackageCoverage, filename="cov.xml")
...
+
+If `test_args` are provided, the top level `"total"` key is replaced with the name of the test set.
+
+Args:
+    coverage (PackageCoverage): Coverage metrics evaluated for the package.
+    filename (str, optional): Filename of the generated JSON summary. Defaults to "coverage-summary.json".
+    test_args (list of str, optional): Arguments passed to `Pkg.test`. Defaults to `[""]`.
```
Comment (tpapp) [inline src/LocalCoverage.jl]:
```suggestion
    - `coverage` (PackageCoverage): Coverage metrics evaluated for the package.
    - `filename` (str, optional): Filename of the generated JSON summary. Defaults to "coverage-summary.json".
    - `test_args` (list of str, optional): Arguments passed to `Pkg.test`. Defaults to `[""]`.
```

File: src/LocalCoverage.jl [hunk 8]
```diff
@@ -449,6 +482,92 @@ function generate_xml(coverage::PackageCoverage, filename="cov.xml")
...
+
+Returns:
+    str: The absolute path to the generated JSON summary file.
```
Comment (tpapp) [inline src/LocalCoverage.jl]:
```suggestion
    The absolute path to the generated JSON summary file.
```
Return values need no name.

File: src/LocalCoverage.jl [hunk 9]
```diff
@@ -449,6 +482,92 @@ function generate_xml(coverage::PackageCoverage, filename="cov.xml")
...
+
+JSON Schema:
+    The output JSON file maps each source file (and a package `"total"` or custom test set name)
+    to a set of coverage metrics (`lines`, `statements`, `functions`, and `branches`):
+
+    {
```
Comment (tpapp) [inline src/LocalCoverage.jl:506]: should probably quote this
````suggestion
```
    {
````

File: src/LocalCoverage.jl [hunk 10]
```diff
@@ -449,6 +482,92 @@ function generate_xml(coverage::PackageCoverage, filename="cov.xml")
...
+            "branches": {"total": int, "covered": int, "skipped": int, "pct": float}
+        },
+        "src/file.jl": {
+            "lines": {"total": int, "covered": int, "skipped": int, "pct": float},
+            "statements": {"total": int, "covered": int, "skipped": int, "pct": float},
+            "functions": {"total": int, "covered": int, "skipped": int, "pct": float},
+            "branches": {"total": int, "covered": int, "skipped": int, "pct": float}
+        }
+    }
+
```
Comment (tpapp) [inline src/LocalCoverage.jl]: and close with ` ``` ` here

**Review by tpapp — CHANGES_REQUESTED** [review]
Thanks, very nicely organized code! just asked for a few cosmetic changes, then this is good to go.

File: src/LocalCoverage.jl (hunk 8 above)
Comment (twilsonco) [reply src/LocalCoverage.jl]: That's the return type, not a name. Still want it removed?

Comment (tpapp) [reply src/LocalCoverage.jl]: then please spell it out (`AbstractString`?)

Comment (twilsonco) [reply src/LocalCoverage.jl]: Ah! My Python was showing! Thanks for spelling *that* out for me.

**Review by tpapp — APPROVED** [review]
LGTM, thanks for the persistence and the nice work!
